documentation/design-docs/ipc-protocol.md

```diff
@@ -0,0 +1,183 @@
+# Event Pipe
+
+## Overview
+
+> TODO: Write an overview of what this does
+
+## TODOs
+* formalize struct names
+* formalize error codes (Use HRESULT values?)
+
+## Inter-Process Communication Specification
+
+### Current
+
+```c++
+enum class DiagnosticMessageType : uint32_t
+{
+    // EventPipe
+    StartEventPipeTracing = 1024, // To file
+    StopEventPipeTracing,
+    CollectEventPipeTracing, // To IPC
+};
+
+struct MessageHeader
+{
+    DiagnosticMessageType RequestType;
+    uint32_t Pid;
+};
+```
+
+```
+runtime <- client : MessageHeader { CollectEventPipeTracing }
+    error? -> no response
+runtime -> client : session ID 
+runtime -> client : event stream
+
+...
+
+runtime <- client : stop command
+runtime -> client : session id and stops
+```
+
+### Proposed
+
+Connection:
+```
+runtime <- client : handshake (set length byte array that represents version)
+runtime -> client : sends handshake back as ACK
+runtime <- client : packet containing start code
+runtime -> client : ACK + potential payload
+repeat till closed
```

Comment:
> If we decide to use this protocol for communicating during a debugging session

The debugger is a more complicated case, and I think it is very likely we'll leave it alone for a long time to come. I think the best option for now is to consider it out of scope.

> In your mind, is the command/control connection transient, i.e., I send a command and if I send another I'm initiating another connection?

Yeah, or at least I'd treat it as the initial strawman because it is what we already implemented and it is simple. If we want to do something different than that I think we should understand what extra value are we getting that justifies adding the change and complexity. For example HTTP started as single command per connection and then eventually added connection re-use. In their case they did it for performance, but I don't expect performance to be compelling here because we don't need high volume of commands.

Reply:
Okay, that makes sense.  I think having the handshake as a separate interaction gives us the ability to eventually have persistent command/control sessions, while not locking us out of the current interaction.  It also allows us to keep the packet header quite small.  The alternative would be to put the magic version number in the packet header which would increase the size by however many bytes it takes to represent that number.

I think the best way spec this would be to say: "On initial connection, the handshake will be exchanged, and all subsequent commands over _that same_ transport do not require the handshake"  This allows us to have the concept of persistent sessions at a later date without having to do work now to support them.  Thoughts?

Reply:
If the protocol says the client is allowed to do something (ie send a 2nd command) and the runtime says it implements the protocol I would interpret that as the runtime has to support it now. You could use a new handshake in the future (DOTNET_IPC_V2) to indicate that now it is legal to reuse the connection for multiple commands. The client sends the handshake and either the runtime accepts it if it supports that, or sends an error indicating it isn't going to support that.

Reply:
I'm coming around to your way of thinking on this.  I can modify the header to contain a much shorter magic version number:
```c
struct header
{
    char[8] magic = { 0x44 ,0x4E ,0x49 ,0x50 ,0x43 ,0x31, 0x00, 0x00 }; // "DNIPC1"
    uint16_t size = 17; // header alone is 17 bytes
    uint16_t command = 0x0101; // combined command field
    char reserved = 0;
}
```

The read code would look something along the lines of 
```c
struct header header;
int n_read = read(sizeof(header)), &header);
// error checking on read

switch (*(uint64_t*)&header.magic) //or we just pass it as a uint64 in the first place...
{
    case IPCVersions.DNIPC1:
        // hand off packet to v1 handler
        break;
    ...
}
```